Added Masking of PII data in logs

SuyeshBadge opened the pull request:
## Pull Request Template
Added feature to mask PII data into logs.

**Description:**

- This PR introduces a new feature to mask Personally Identifiable Information (PII) fields in log entries. The `MaskingFilter` in the `logging` package has been updated to support masking of various PII fields, such as name, email, phone number, Social Security number, credit card number, date of birth, biometric data, and IP address.
- The motivation behind this change is to enhance data privacy and security by preventing sensitive user information from being logged in plaintext.
- This feature will benefit users by ensuring compliance with privacy regulations and reducing the risk of data breaches or unauthorized access to sensitive information.
- Closes: https://github.com/gofr-dev/gofr/issues/459

**Breaking Changes (if applicable):**

- There are no breaking changes introduced by this PR.

**Additional Information:**

- The changes rely on the built-in `reflect` package in Go for inspecting and modifying struct fields.
- New test cases have been added to ensure comprehensive testing of the PII masking functionality.

**Checklist:**

- [x] I have formatted my code using `goimport` and `golangci-lint`.
- [x] All new code is covered by unit tests.
- [x] This PR does not decrease the overall code coverage.
- [x] I have reviewed the code comments and documentation for clarity.

**Thank you for your contribution!**

SuyeshBadge commented:
@aryanmehrotra, @vipul-rawat, and @Umang01-hash,

I'm initiating a discussion thread to commence the implementation of the mentioned feature. Let's delve into the various potential changes and methodologies to effectively incorporate this feature.

Looking forward to your valuable insights and contributions.

vipul-rawat commented on pkg/gofr/logging/logger.go:
```diff
@@ -223,24 +297,69 @@ func colorForGRPCCode(status int32) int {
+func NewLogger(level Level, args ...interface{}) Logger {
+	var filter Filter
+	if len(args) > 0 {
+		filter = args[0].(Filter)
```
This would panic if the args provided do not implement `Filter` interface.
We can do a boolean(ok) check instead.

vipul-rawat commented on pkg/gofr/logging/logger.go:
```diff
@@ -223,24 +297,69 @@ func colorForGRPCCode(status int32) int {
+
+	var filter Filter
+	if len(args) > 0 {
+		filter = args[0].(Filter)
```
This would panic if the args provided do not implement `Filter` interface.
We can do a boolean(ok) check instead.

vipul-rawat commented on pkg/gofr/logging/logger_test.go:
```diff
@@ -387,3 +387,78 @@ func Test_clean(t *testing.T) {
+			},
+			expected: `{"phoneNumber":0,"socialSecurityNumber":0,"creditCardNumber":0,"dateOfBirth":0,"biometricData":0,"ipAddress":"***********"}`,
+		},
+		// Add more test cases as needed
```
Empty comment

vipul-rawat commented on pkg/gofr/logging/logger.go:
```diff
@@ -41,11 +42,81 @@ type Logger interface {
+	return entry
+}
+
+func contains(slice []string, str string) bool {
```
naming convention could be better for `slice`, not informative of the usage

vipul-rawat reviewed (commented):
Please check the linters, pipeline is failing.

Also is there a way to not use reflect, if possible?

You might also need to look for a way for users to the masking fields at the time of application creation! Or if masking is required at all!

SuyeshBadge commented:
Hey @vipul-rawat, I've pushed the requested changes along with the formatting and linting fixes. Please have a look once you get a chance and let me know if anything is missing.

vipul-rawat reviewed (changes requested):
There are still 2 things - 
1. Can you add an example on how enabling and disabling the filtered logging would look from user's perspective
2. I am still not sure if this would handle the pointers

SuyeshBadge commented:
Certainly, @vipul-rawat will make the changes and update you accordingly.

SuyeshBadge commented:
> I am still not sure if this would handle the pointers

You're right to be concerned about handling pointers. The current implementation assumes that the fields are direct values and not pointers. Will update this too.

SuyeshBadge commented:
Hey @vipul-rawat ,
I've updated the filtered logging feature to support pointer values and added documentation on how to use it. However, one test check is failing, and I'm not sure why.
Could you please take a look and provide some guidance on fixing it?
Thanks for your help!

vipul-rawat commented:
@SuyeshBadge This implementation requires users to be aware of the logger and define it themselves, whereas for most use cases the framework is responsible for managing the logger itself. Can we use configs reading to achieve the same instead of the user defining the filter in the main file?

Also, how can a user-defined filter be called `DefaultFilter`? Should be CustomFilter or equivalent.

The pipeline was failing due to an MQTT test! Probably the public server was experiencing some downtime. Rerunning should fix the same. Also, update the local with development.

SuyeshBadge commented:
Hello @vipul-rawat,

Could you kindly provide further clarification regarding the utilization of configuration readings in this context? Specifically, would you prefer me to:

1. Pass the configuration explicitly to the `NewLogger` function?
2. Expose a `setConfig` method that will be invoked prior to declaring the logger?
3. Directly invoke `godotenv` from the `logger.go` file?

I require additional details to proceed effectively. Thank you.

vipul-rawat commented:
@SuyeshBadge What I mean is that currently according to your implementation, the NewLogger is being made by the user in main, but what we want is that the filter gets configured automatically during the app creation (during `gofr.New()`), so for example if user gives the config `MASKED_FIELDS=<comma separated field names>` then what framework does is that it adds the filter with those names, we don't want to have the filtering of fields by default. 

I think something like `SetFilter` similar to `setConfig` that is suggested in point 2 seems good.

> Point 1 will not be viable as it would be a breaking change and not worth it.
> Point 3 also not viable as logger's responsibility is not to read configurations or load them.

@srijan-27 FYA

SuyeshBadge commented:
Hey @vipul-rawat,

I've implemented the changes as per your instructions. Additionally, I addressed several linting issues that were present. Kindly review the updates at your convenience. With the latest implementation, the application now reads values from the configuration and configures the logger accordingly. Furthermore, I've revised the documentation for improved clarity and understanding.

SuyeshBadge commented:
Also if possible could you please re-run the checks, some of them are failing.

srijan-27 commented on docs/advanced-guide/filtered-logging/page.md:
```diff
@@ -0,0 +1,72 @@
+
+To use filtered logging, you need to set the masking configuration in your `.env` file. The following configuration options are available:
+
+- `LOGGER_MASKING_ENABLED`: Set to `true` to enable masking of sensitive fields in log messages. Default is `false`.
+- `LOGGER_MASKING_FIELDS`: A comma-separated list of field names that should be masked in log messages.
```
@SuyeshBadge - This use-case can be fulfilled with a single config, not required to create a bool config as well. Please fix this.

srijan-27 commented on docs/advanced-guide/filtered-logging/page.md:
```diff
@@ -0,0 +1,72 @@
+    app := gofr.New()
+
+    // Use the logger
+    logger := logging.NewLogger(logging.INFO)
```
it is not required to create a new logger. We can use `app.Logger()`

srijan-27 commented on pkg/gofr/container/container.go:
```diff
@@ -54,80 +54,126 @@ func NewContainer(conf config.Config) *Container {
 }
 
 func (c *Container) Create(conf config.Config) {
+	c.setAppName(conf)
+	c.setAppVersion(conf)
+	c.createLogger(conf)
+	c.createMetricsManager()
+	c.createRedis(conf)
+	c.createSQL(conf)
+	c.createPubSub(conf)
```
This should not be included in this PR, as it is not directly related to masking logs. Please create a separate PR for this.

srijan-27 commented on pkg/gofr/container/container.go:
```diff
@@ -54,80 +54,126 @@ func NewContainer(conf config.Config) *Container {
 
+func (c *Container) createMetricsManager() {
 	c.metricsManager = metrics.NewMetricsManager(exporters.Prometheus(c.appName, c.appVersion), c.Logger)
-
-	// Register framework metrics
 	c.registerFrameworkMetrics()
-
-	// Populating an instance of app_info with the app details, the value is set as 1 to depict the no. of instances
```
deleting the comments is not required. Please revert

srijan-27 commented:
@SuyeshBadge - Are you still working on this?

SuyeshBadge commented:
> @SuyeshBadge - Are you still working on this?

Yes @srijan-27 will update soon

SuyeshBadge commented:
Hey @srijan-27 and @vipul-rawat,

The changes we discussed have been implemented. Please review them and let me know if anything else is needed.

Thanks!

srijan-27 commented on pkg/gofr/logging/logger_test.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"encoding/json"
 	"fmt"
+	"reflect"
+
```
unwanted extra newline, it can be removed.

srijan-27 commented on pkg/gofr/logging/logger_test.go:
```diff
@@ -214,3 +216,284 @@ func TestPrettyPrint(t *testing.T) {
+		return
+	}
+
+	expected := `{"name":"********","email":"********************","password":"**********"}`
+	if string(maskedMessage) != expected {
+		t.Errorf("Expected %s, but got %s", expected, maskedMessage)
+	}
+
+	if !reflect.DeepEqual(filteredMessage, expectedOutput) {
+		t.Errorf("Expected output %v, but got %v", expectedOutput, filteredMessage)
+	}
```
same as above, assert.Equal can be used.

srijan-27 reviewed (changes requested):
@SuyeshBadge - please take a look

SuyeshBadge commented:
Hey @srijan-27 @vipul-rawat,

I see there have been some recent changes pushed on the PR. Is there anything pending on my end? If yes, please let me know. Otherwise, let's move forward with this.

Thanks!

srijan-27 commented:
@SuyeshBadge - Have you tested these changes by running an example on local and verifying whether the fields of struct are getting masked?

SuyeshBadge commented:
Not yet @srijan-27 .

aryanmehrotra commented:
@SuyeshBadge can we connect and close this ASAP?
If you are stuck we can work together!

srijan-27 commented:
> Certainly, @aryanmehrotra .
> 
> I identified the issue with the filter not working and have implemented a fix. However, we can still connect if you'd like to discuss the changes in detail.

@SuyeshBadge - If you have implemented the fix, please push the code and we can verify then.

SuyeshBadge commented:
Sure folks, will push it today.

SuyeshBadge commented:
Hey Folks, Pushed some latest changes please take a look and let me know if this works. 

FYI @srijan-27 @aryanmehrotra

srijan-27 commented:
@SuyeshBadge - can you please attach the snapshots for the masked logs, if you have tested it locally on any example.

SuyeshBadge commented:
Sure, @srijan-27. Attached are the snapshots of the logs, the code, and the keys that need to be masked.

- **Snapshot of fields that need to be masked**  
  ![fields to be masked](https://github.com/gofr-dev/gofr/assets/59205766/f7d39630-b716-47a4-896d-93118bde02fe)

- **Snapshot of the log after masking**  
  ![log of masking](https://github.com/gofr-dev/gofr/assets/59205766/607b455c-b8dd-48b6-b148-819bee5e8943)

- **Snapshot of logs without masking**  
  ![logs without masking](https://github.com/gofr-dev/gofr/assets/59205766/bd0828e4-c466-4df0-82af-41664847cdc2)

- **Snapshot of the code snippet**  
  ![code for masking](https://github.com/gofr-dev/gofr/assets/59205766/7abb8871-503b-42ab-8254-177fe26d47dd)

SuyeshBadge commented:
Sure, @srijan-27. Attached are the snapshots of the logs, the code, and the keys that need to be masked.

- **Snapshot of fields that need to be masked**  
  ![fields to be masked](https://github.com/gofr-dev/gofr/assets/59205766/f7d39630-b716-47a4-896d-93118bde02fe)

- **Snapshot of the log after masking**  
  ![log of masking](https://github.com/gofr-dev/gofr/assets/59205766/607b455c-b8dd-48b6-b148-819bee5e8943)

- **Snapshot of logs without masking**  
  ![logs without masking](https://github.com/gofr-dev/gofr/assets/59205766/bd0828e4-c466-4df0-82af-41664847cdc2)

- **Snapshot of the code snippet**  
  ![code for masking](https://github.com/gofr-dev/gofr/assets/59205766/7abb8871-503b-42ab-8254-177fe26d47dd)

- **Snapshot of the log, if the data passed, is a pointer**
![code](https://github.com/gofr-dev/gofr/assets/59205766/eb60f6ab-de15-4b06-a106-1754cb2498c9)

![Pointer](https://github.com/gofr-dev/gofr/assets/59205766/a64cdcd8-e83c-4daa-b975-829d2eae8f55)

srijan-27 commented:
@SuyeshBadge - have you tried passing the struct as args in log?

Something like this:
```go
func helloHandler(ctx *gofr.Context) (interface{}, error) {
	user := struct {
		Name     string
		Email    string
		Password string
	}{
		"srijan", "srijan.rastogi@gofr.dev", "secret-123",
	}

	ctx.Logger.Infof("%v", user)

	return "hello world!", nil
}
```

SuyeshBadge commented:
Hey @srijan-27,
In this case, the output will not be masked because the message being passed to the filter method is a string that is formatted before being passed. I'm not entirely sure if this is the intended behavior. Let me know what you think and what we can do here.

srijan-27 commented:
@SuyeshBadge - I don't think this should be intended behaviour. As if we consider this, then we are restricting user to only mask fields in unformatted logs. We should have a behaviour such that user can mask fields irrespective of the log formatting.

SuyeshBadge commented:
Hey @srijan-27, Can we discuss this over a call? Let me know if you are available.

SuyeshBadge commented:
Hi @srijan-27,

The issue is fixed. We are now masking the arguments, ensuring that all types of data passed are masked and not limited to any specific case. If you encounter any more edge cases, please feel free to raise them.

Also, let me know if this solution works for you. I am attaching a snapshot of the updated use cases.

---
**Snapshots**
![Struct example](https://github.com/gofr-dev/gofr/assets/59205766/4708d92e-0164-40aa-8402-a34426db39db)
![terminal](https://github.com/gofr-dev/gofr/assets/59205766/96262829-ecd8-4584-8890-6a19fb604606)

srijan-27 commented:
@SuyeshBadge - Although this implementation is masking the fields as required. But it is also adding a log which is not helpful and not understandable for user. See below:
<img width="1358" alt="Screenshot 2024-06-20 at 12 15 11 PM" src="https://github.com/gofr-dev/gofr/assets/44723623/bc2d9c1d-7ea0-4b04-830f-1f4ab092dcf7">

Also, the request/response log is missing in the new implementation:
<img width="621" alt="Screenshot 2024-06-20 at 12 15 40 PM" src="https://github.com/gofr-dev/gofr/assets/44723623/3e9c7780-820e-4554-b29d-99c1b058808d">

Please check this.

SuyeshBadge commented:
> @SuyeshBadge - Although this implementation is masking the fields as required. But it is also adding a log which is not helpful and not understandable for user. See below: <img alt="Screenshot 2024-06-20 at 12 15 11 PM" width="1358" src="https://private-user-images.githubusercontent.com/44723623/341301382-bc2d9c1d-7ea0-4b04-830f-1f4ab092dcf7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NDA0ODgsIm5iZiI6MTcxODk0MDE4OCwicGF0aCI6Ii80NDcyMzYyMy8zNDEzMDEzODItYmMyZDljMWQtN2VhMC00YjA0LTgzMGYtMWY0YWIwOTJkY2Y3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDAzMjMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkNzI2MDQyMzA3MzhjOTI2MzUzYjA2NDRjZDFkYzc1MDdiYjcxMmRlZTYxYTU5ODBhZDM5MDNlZWE3Yjk0NjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6zzhU27U0ewfpg36ZIJFBJnR2KcMJvRKiNMGGB_xZ14">
> 
> Also, the request/response log is missing in the new implementation: <img alt="Screenshot 2024-06-20 at 12 15 40 PM" width="621" src="https://private-user-images.githubusercontent.com/44723623/341301961-3e9c7780-820e-4554-b29d-99c1b058808d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5NDA0ODgsIm5iZiI6MTcxODk0MDE4OCwicGF0aCI6Ii80NDcyMzYyMy8zNDEzMDE5NjEtM2U5Yzc3ODAtODIwZS00NTU0LWIyOWQtOTljMWIwNTg4MDhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIxVDAzMjMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3MjYwMWY2MDhlNGQxODNjMDAxY2E3ZGI5ZjdmMzBlOWY2OGEzM2QwZjk3ZDEzNGU3OGRmZTFlOTNiODBmMmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._mbfUyfJirsxe6tbTc5I_QNgWFqL0VYnpzW3m_biEnA">
> 
> Please check this.

Ohh Let me check why this log is being added here.

srijan-27 commented:
Spent time on fixing this implementation, but seems like a lot of issues with the current implementation like: not proper masking, breaking on existing logs. And also since this is using reflections a lot, so the performance is affected considerably. 
Hence, closing this PR. Would need to think of a better solution.
JFYI: @vipul-rawat, @aryanmehrotra, @Umang01-hash

srinivas365 commented:
@srijan-27
I have created a package to redact the pii data. Can I implement this in gofr?
https://github.com/srinivas365/piishield/tree/v1.0.0

srijan-27 commented:
@srinivas365 - Sure, you can work on this issue: #459.
Please drop a comment there so I can assign this to you.
